pkg/cri/resource-manager/policy/builtin/memtier/pod-preferences.go

```diff
@@ -217,62 +325,123 @@ func coldStartPreference(pod cache.Pod, container cache.Container) (ColdStartPre
+	//      - 1 full core < CPU request
+	//          => fractional/shared cores
+	//      - 1 full core <= CPU request < 2 full cores:
+	//          a. fractional allocation:
+	//            - shared preference explicitly annotated/configured false:
+	//              => mixed cores, prefer isolated, unless annotated/configured otherwise
+	//            - shared preference explicitly annotated/configured true:
+	//              => shared cores
+	//          b. non-fractional allocation:
+	//            - shared preference explicitly annotated true:
+	//              => shared cores
+	//            - isolated default preference false or explicitly annotated false:
+	//              => exclusive cores
+	//            - isolated default preference true or explicitly annotated true:
+	//              => exclusive cores, prefer isolated
+	//      - 2 full cores <= CPU request
+	//          a. fractional alloction:
+	//            - shared preference explicitly annotated false:
+	//              => mixed cores, prefer isolated only if explicitly annotated
+	//            - otherwise (no shared annotation):
+	//              => shared cores
+	//          b. non-fractional allocation:
+	//            - shared preference explicitly annotated true:
+	//              => shared cores
+	//            - otherwise (no shared annotation):
+	//              => exclusive cores, prefer isolated only if explicitly annotated
+	//
+
+	namespace := container.GetNamespace()
+	request := container.GetResourceRequirements().Requests[corev1.ResourceCPU]
+	qosClass := pod.GetQOSClass()
+	fraction := int(request.MilliValue())
+
+	// easy cases: kube-system namespace, Burstable or BestEffort QoS class containers
+	switch {
+	case namespace == metav1.NamespaceSystem:
+		return 0, fraction, false, cpuReserved
+	case qosClass == corev1.PodQOSBurstable:
+		return 0, fraction, false, cpuNormal
+	case qosClass == corev1.PodQOSBestEffort:
+		return 0, 0, false, cpuNormal
 	}
 
-	qos := pod.GetQOSClass()
+	// complex case: Guaranteed QoS class containers
+	cores := fraction / 1000
+	fraction = fraction % 1000
+	preferIsolated, explicitIsolated := isolatedCPUsPreference(pod, container)
+	preferShared, explicitShared := sharedCPUsPreference(pod, container)
 
-	preferIsol, explicit := podIsolationPreference(pod, container)
-	preferShared, elevate := podSharedCPUPreference(pod, container)
-
-	full, fraction, isolate := 0, 0, false
 	switch {
-	case container.GetNamespace() == metav1.NamespaceSystem:
-		full, fraction = 0, int(req.MilliValue())
-
-	case qos == corev1.PodQOSBurstable || preferShared:
-		full, fraction = 0, int(req.MilliValue())
-
-	case qos == corev1.PodQOSGuaranteed:
-		full = int(req.MilliValue()) / 1000
-		fraction = int(req.MilliValue()) % 1000
-	}
-
-	if !preferShared {
-		switch {
-		case full == 1:
-			if explicit {
-				isolate = preferIsol
-			} else {
-				isolate = true
+	// sub-core CPU request
+	case cores == 0 && fraction < 1000:
```

Comment:
`fraction < 1000` is always true after line 392 (`fraction = fraction % 1000`)

Reply:
Thanks, true. Removed tautological comparison.